Interest System

netcables opened the pull request:
With these changes, a user is able to add and remove user and conversation interests while signed in by using the commands `i-u-add <name>`, `i-u-remove <name>`, `i-c-add <title>`, and `i-c-remove <title>`. Interests are properly stored on the server and remain even if a user signs off and back on.

Commands for status updates are `status-update-u <name>` and `status-update-c `<title>`. When a status update is called for a conversation, the amount of new messages since the last status update on it is returned if the conversation is in the user's interests. When a status update is called for a user, conversations that they have created and contributed to since the last status update on them are returned if the user is in the user's interests.

netcables committed:
Added constants for Interest System

Added constants for User Interest, Conversation Interest, and Status Update.

netcables committed:
Added handling of Interest System requests

Added handling requests for Add User Interest, Add Conversation Interest, and Status Update.

netcables committed:
Modified request handling

Changed how the server handles interest-related requests

netcables committed:
Added placeholder commands

Added placeholder commands for adding users and conversations to interests.

netcables committed:
Added more placeholder commands

Added the placeholder commands to remove interests. Also updated existing placeholders.

netcables committed:
Started work on serializers

Started work on keeping interests tracked by the server.

netcables committed:
Additions to Interest System

Moved methods from View to Controller and modified how Server handles requests related to Interest System.

netcables committed:
Added Interest System methods

Added Interest System methods to Controller and Server

netcables committed:
Removed unused loops

Removed the Uuid for loops in User.

netcables committed:
Fixed error involving User Interests

Fixed a bug that caused the client to send the wrong UUID when making a request to add or remove user interests.

netcables committed:
Added more log information

Changed Controller to show user and interest names and UUIDs when methods related to Interest System are used.

netcables committed:
Fixed Interest Removal

Fixed an error in which the methods related to conversation interests were calling methods related to user interests.

netcables committed:
Removed casting

Added methods to BasicController to avoid casting.

netcables committed:
Added client-side messages

The client is now properly notified when interest changes take place server-side.

netcables committed:
Added more comments

Added comments to describe Interest System methods

netcables committed:
Added placeholders for status update

Added network codes and placeholder methods for status update.

netcables committed:
Added status updated methods to View

Added userStatusUpdate and conversationStatusUpdate placeholders to view.

netcables committed:
Fixed return issue in View

Fixed View missing a return statement by making the method conversationStatusUpdate return nothing.

netcables commented:
I'm aware of the strange indentation in a few areas. Working on cleaning it up right now.

netcables committed:
Fixed indentation

Fixed inconsistent indentation in Controller.

evelez7 commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -328,6 +340,166 @@ private ConversationContext find(String title) {
+            final ConversationContext conversation = find(title);
+            if (conversation == null) {
+              System.out.format("ERROR: No conversation with name '%s'\n", title);
+            } else {
```
An `else if` block can be used here to condense the code and avoid creating another nested `if` block. This change should be applied to the applicable `if` blocks in the rest of the methods (like i-c-remove and i-u-add).

evelez7 commented on src/codeu/chat/server/Controller.java:
```diff
@@ -143,6 +143,78 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
+      LOG.info("User Interest added: " + foundUser.id);
+      return true;
+    }
+    
```
Nit: Eliminate white space between brackets to keep consistency

evelez7 commented on src/codeu/chat/server/Controller.java:
```diff
@@ -143,6 +143,78 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
+      LOG.info("ERROR: User already in interests.");
+      return false;
+    }
+    else {
```
Nit: Keep `else` next to closing bracket from `if` blocks to keep consistency.

evelez7 reviewed (commented):
Nested if statements can be a little condensed. A couple nits but otherwise LGTM.

netcables committed:
Fixed consistency issues

Cleaned up a couple of areas that had extra white space or had "else" taking up entire lines. Also condensed a few spots by using "else if".

netcables committed:
Added more placeholders for status update

Added a few more placeholders for conversationStatusUpdate.

netcables committed:
Added user status update placeholders

Added placeholders for implementation of user status update.

netcables committed:
Updated conversation status update

Conversation status update is now working properly. The code is still currently in the process of being cleaned up.

netcables committed:
Used HashMaps for updates

Switched from a single last updated value for users and conversations to HashMaps in order to track updates per individual user or conversation

netcables committed:
Updated user status update

User status update is now working properly. As with conversation status update, everything still needs to be cleaned up.

netcables committed:
Fixed user status update

Added a missing check for if a specified user recently created a conversation to user status update.

evelez7 reviewed (commented):
LGTM

netcables committed:
Removed copies of "find" method

Switched interest system methods from returning boolean values to returning integer values in order to remove the multiple copies of a "find" method in Chat by checking if the specified user or conversation exists on the server side rather than the client side.

netcables committed:
Removed another "find" method copy

Removed the copy of the "find" method used by status-update-c. Working toward removing the last copy present in status-update-u.

netcables committed:
Removed final copy of "find" method

Removed the last remaining copy of "find" in Chat and returned UserPanel to its original state (it no longer takes Context as an argument).

netcables commented:
I'm having some issues in Chat.java regarding Time. User Status Update was working perfectly fine several days ago, but for some reason, the last status update time for User Status Update always seems to be set to the current time (which prevents status updates on users from working properly). I'm using the same Time.now() method that I'm using for Conversation Status Update, and that still works fine.

Could you guys please be on the lookout for any mistakes that I made that could have caused this?

Tuxdude commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -243,6 +243,18 @@ public void invoke(List<String> args) {
+        System.out.println("  i-c-remove <title>");
+        System.out.println("    Remove a conversation from the current user's interests.");
+        System.out.println("  status-update-u <name>");
+        System.out.println("    Call a status update on the specified user interest.");
```
nit: `Request status update` here and in the next command description too.

Tuxdude commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -328,6 +340,166 @@ private ConversationContext find(String title) {
+      @Override
+      public void invoke(List<String> args) {
+        for(String token : args){
+          final String title = token;
```
nit: This line is redundant. You can use just `token` or `title` everywhere and avoid this assignment.

Please also fix this in other places.

Tuxdude commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -328,6 +340,166 @@ private ConversationContext find(String title) {
+        for(String token : args){
+          final String title = token;
+          if (title.length() > 0) {
+            int response = user.addConversationInterest(title);
```
Instead of returning an `int` response code for `addConversationInterest`, it will be more readable if you returned an enum which said `NO_ERRROR`, `ERROR_FAILED_DUE_TO_SOMETHING`, `ERROR_FAILED_DUE_TO_SOMETHING_ELSE`, etc.

Also, the general convention is to use `0` to indicate no error, and non-zero values to indicate some errors. Even then, enums would be more readable than returning integers for such cases.

Please also fix this in other places.

Tuxdude commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -328,6 +340,166 @@ private ConversationContext find(String title) {
+          final String title = token;
+          if (title.length() > 0) {
+            int response = user.addConversationInterest(title);
+            if (response == 1) {
```
Both `int` and `enum` can use switch-case instead of long `if ... else if ...` blocks.

Please fix this in other places too.

Tuxdude commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -328,6 +340,166 @@ private ConversationContext find(String title) {
+          if (name.length() > 0) {
+            final ArrayList<String> contributions = user.userStatusUpdate(name);
+            if (contributions.isEmpty()) {
+              System.out.format("ERROR: No user with name '%s'\n", name);
```
Could this also indicate the user did not contribute to any conversations ? If so, you might just want to update the printed message.

Tuxdude commented on src/codeu/chat/client/core/Controller.java:
```diff
@@ -112,4 +112,105 @@ public ConversationHeader newConversation(String title, Uuid owner)  {
+    try (final Connection connection = source.connect()) {
+
+      Serializers.INTEGER.write(connection.out(), NetworkCode.NEW_CONVERSATION_INTEREST_REQUEST);
+      Serializers.STRING.write(connection.out(), title);
```
Are conversation titles unique? What will happen if there are 2 conversations with the same title?

netcables replied:
Conversation titles and usernames actually aren't unique. In both cases, the one that was created first is the one that gets selected any time that the model searches by title/name (this is also the case when signing in as a user that has the same name as another user). Do you think that creating another version of this commands that take UUIDs as an argument would be the best solution for this? We originally planned to keep it simple so that users wouldn't have to type out long UUIDs every time that they wanted to change interests.

Tuxdude commented on src/codeu/chat/client/core/UserContext.java:
```diff
@@ -53,4 +53,33 @@ public ConversationContext start(String name) {
+
+  public ArrayList<String> userStatusUpdate(String name) {
+    final ArrayList<String> contributions = new ArrayList<>();
+    for(final String contribution : view.userStatusUpdate(name, user.id)) {
```
Could we just return the result of `view.userStatusUpdate(name, user.id);` or am I missing anything ?

netcables replied:
Just fixed this. I forgot that I meant to import Collection here and just return a Collection rather than moving everything into an ArrayList and returning that (because ArrayList was already imported).

Tuxdude reviewed (commented):
Adding my initial set of comments. I might need to do another pass at the code, especially `View.java`.

For the most part, the code looks good.

Please consider adding some basic unit tests to validate the logic, since many things might be easily get overlooked in code reviews (including my reviews).

netcables committed:
Replaced long 'if else' blocks

Replaced the if else blocks checking values with switches and also made interest system methods in UserContext return enum values instead of ints.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,81 @@ public View(Model model) {
+    final User foundUser = model.userByText().first(name);
+
+    if(foundUser != null) {
+      if(foundOwner.UserSet.contains(foundUser.id)) {
```
Can `foundOwner` be `null` here? Do you want to verify it is not null ?

netcables replied:
foundOwner is null when a user with the specified name cannot be found in Model. The check in there is to prevent the following check from having errors (because it would check the current user's UserSet for a UUID from a User object that doesn't exist).

Tuxdude replied:
I think your comment talks about `foundUser` which makes sense and is a good check to have.

What I'm referring to is a similar check you're currently missing for `foundOwner`. If that happens to be `null`, you will get a `NullPointerException`.

netcables replied:
Oh, whoops. Meant to say foundUser. I'm pretty sure that foundOwner can't be null in this case because the UUID provided as the argument is the UUID of the user that is currently signed in. It's impossible to sign in as a user if they're not in Model (and if they are, they have a UUID.

Tuxdude replied:
IIUC, you accept the UUID from the incoming message (correct me if I'm wrong here). What will happen if the incoming message contains a garbage UUID? I understand the current client will not send it, but what if someone writes a client which does?

This is not a big concern for the scope of this project, but definitely a thing to keep in mind. It is better to at least throw an exception or error for such cases rather than crashing with a RuntimeException (like NullPointerException).

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,81 @@ public View(Model model) {
   public Message findMessage(Uuid id) { return model.messageById().first(id); }
 
+  @Override
+  public Collection<String> userStatusUpdate(String name, Uuid owner) {
```
This function and `conversationStatusUpdate` are both nested quite a bit (few if blocks and for loops). Generally anything more than 2 nested blocks, it is always easier if you split them into smaller methods (you can keep them private if you do not want to expose them outside).

Tuxdude replied:
Also adding some comments here would be useful. It is slightly difficult to follow the code otherwise.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,81 @@ public View(Model model) {
+            if(lastUpdate.compareTo(currentMessage.creation) < 0 && currentMessage.author.equals(foundUser.id)) {
+              ConversationHeader conversationContribution = model.conversationById().first(conversationPayload.id);
+              contributions.add(conversationContribution.title);
+              LOG.info("Conversation added to list.");
```
Might be a good idea to make this log message `verbose`.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,81 @@ public View(Model model) {
+              ConversationHeader conversationContribution = model.conversationById().first(conversationPayload.id);
+              contributions.add(conversationContribution.title);
+              LOG.info("Conversation added to list.");
+              foundMessage = true;
```
Why do we set foundMessage to true here and exit the loop? What if there are more messages of interest here ? We don't publish those to the user?

netcables replied:
status-update-u <name> is meant to give the user that requests it a list of conversations that the person they specify has recently contributed to. I just kind of assumed that if at least one message from the specified person is in the conversation and had a creation time AFTER the last time a status update was called on them, they contributed to the conversation.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,81 @@ public View(Model model) {
+            }
+            currentMessage = model.messageById().first(currentMessage.next);
+          }
+          if(!contributions.contains(model.conversationById().first(conversationPayload.id).title)) {
```
I'm sorry but these 3 if blocks are not readable, especially I did not follow why we have this check in addition to the above for loop. I'm sure I'm missing something.

Please add some comments and may be move them to a helper function for better readability.

netcables replied:
The first check is to see if the user exists (i.e. the specified user that was named was added to the server at some point with u-add <name> and can be found in Model).

netcables replied:
The second check is to see if the user is in the current user's user interest set (i.e the specified user that was named was added to the current signed in user's interests at some point with i-u-add and can be found in user.UserSet). A user can only call status updates on users that they have added to their user interest set.

netcables replied:
The for loop goes through all of the ConversationPayloads stored in Model.

netcables replied:
The while loop goes through all of the messages in the current ConversationPayload and checks to see if any of the messages' author UUID matches the UUID of the specified user and if the creation time is after the last time the current user has requested a status update on the specified user. If so, the conversation's title is added to the list and the loop is stopped before it reaches the last message of the conversation.

netcables replied:
The final if statement checks if the current conversation's title was added to the list (through the previous while loop). If it was not, it checks to see if the conversation's creator matches the specified user and if the conversation's creation time is after the last status update request. If both of these are true, the conversation is added to the list and "(Creator)" is added to the string to note that the specified user created this conversation recently.

Tuxdude replied:
I meant comments in the code :) You can just move these into the code, so that it would be easier to read.

Tuxdude reviewed (commented):
Mostly readability related. Please fix these and ping me. I can take a look at `View.java` after that. Thanks!

netcables committed:
Added comments for Interest System

Added comments to explain User Status Update and Conversation Status Update.

netcables committed:
Changed returned int values

Adding and removing interests now return 0 when an interest is successfully modified. -1 is returned in the case where the specified interest is already added or removed and -2 is returned when the specified interest cannot be found in Model.

netcables committed:
userStatusUpdate now returns Collection

Changed userStatusUpdate to return a Collection of strings rather than an ArrayList of strings in order to prevent having to move everything from the Collection returned by view.userStatusUpdate into an ArrayList before returning the information to Chat.

netcables committed:
Added helper functions

Added helper functions for userStatusUpdate and conversationStatusUpdate as well as more comments to describe what each does.

netcables commented on src/codeu/chat/server/View.java:
```diff
@@ -78,6 +78,140 @@ public View(Model model) {
+          contributions.add("(No recent conversations)");
+        }
+        // finally, update the time that status update was last requested for the specified user to now
+        foundOwner.UserUpdateMap.put(foundUser.id, Time.now());
```
This one line is causing absolutely all of my problems at the moment. For whatever reason, this executes IMMEDIATELY after `final Time lastUpdate = foundOwner.UserUpdateMap.get(foundUser.id)` and prevents searchContributions from adding any conversation titles to the collection.

netcables replied:
UPDATE: Added one line that added "THIS SHOULD APPEAR BEFORE THE TIME CHANGE" to the log immediately before line 109. When I attempted a user status update, the line was added to the log twice (when the last status update time was first checked, and when it was actually supposed to show up when the time was supposed to be updated). 109 is definitely being executed as soon as line 100 is reached.

netcables replied:
UPDATE AGAIN: Disregard everything that I mentioned about View, error was in Chat. Accidentally called the method twice in a row.

netcables committed:
Changed Time object names

Changed the name of the Time objects that track the last time a status update was requested to reflect whether they are for users or conversations. User Status Update is still broken.

netcables committed:
Removed unused Collection

Removed a Collection of all of the conversations stored in Model that was accidentally left behind (from when userStatusUpdate didn't have a helper function for searching).

netcables committed:
Changed method for updating time

Used "replace" instead of "put" for updating HashMap values to confirm that the time value being overwritten matches the last update time. User Status Update is still not working correctly.

netcables committed:
Fixed HashMap issue

Fixed utilization of "replace" to update status update times. User Status Update still isn't working.

netcables committed:
Reverted to old method for time update

Reverted usage of "replace" to "put" for updating the time at which a status update was last requested.

netcables committed:
User Status Update now working

Fixed a line in Chat that caused User Status Update to be called twice every time the command was used.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -181,6 +149,66 @@ public int conversationStatusUpdate(String title, Uuid owner) {
+      Message currentMessage = model.messageById().first(conversationPayload.firstMessage);
+      boolean foundMessage = false;
+      // go through every message
+      while(currentMessage != null && foundMessage == false) {
```
Can this while loop be made into a separate method? The method will add all the messages to `contributions`.

Tuxdude commented on src/codeu/chat/server/View.java:
```diff
@@ -181,6 +149,66 @@ public int conversationStatusUpdate(String title, Uuid owner) {
+      }
+      // check if the current conversation wasn't added  to the collection after the above loop
+      if(!contributions.contains(model.conversationById().first(conversationPayload.id).title)) {
+        // check if the current conversation's creator matches the specified user
```
Can this be made into a single if statement with multiple conditions joined using `&&` ?

Tuxdude replied:
Also you're using `model.conversationById().first(conversationPayload.id)` few times. Can you assign it to a local variable to improve readability ?

Tuxdude reviewed (commented):
Mostly LGTM. Feel free to make the suggested changes and merge to master.

evelez7 committed:
Readability improvements in searchContributions

evelez7 committed:
Created new method to add to contributions

Tuxdude reviewed (approved):
LGTM. Please consider adding few tests in the future for this code.
